policy-controller/k8s/index/src/index.rs

```diff
@@ -0,0 +1,738 @@
+struct PodPortServer {
+    /// The name of the server resource that matches this port. Unset when no
+    /// server resources match this pod/port (and, i.e., the default policy is
+    /// used).
+    name: Option<String>,
+
+    /// A sender used to broadcast pod port server updates.
+    tx: watch::Sender<InboundServer>,
+
+    /// A receiver that is updated when the pod's server is updated.
+    rx: watch::Receiver<InboundServer>,
+}
+
+/// Holds the state of policy resources for a single namespace.
+#[derive(Debug)]
+struct PolicyIndex {
+    cluster_info: Arc<ClusterInfo>,
+    servers: HashMap<String, server::Server>,
+    server_authorizations: HashMap<String, server_authorization::ServerAuthz>,
+}
+
+// === impl Index ===
+
+impl Index {
+    pub fn shared(cluster_info: impl Into<Arc<ClusterInfo>>) -> SharedIndex {
+        let cluster_info = cluster_info.into();
+        Arc::new(RwLock::new(Self {
+            cluster_info: cluster_info.clone(),
+            namespaces: NamespaceIndex {
+                cluster_info,
+                by_ns: HashMap::new(),
+            },
+        }))
+    }
+
+    /// Obtains a pod:port's server receiver.
+    ///
+    /// An error is returned if the pod is not found. If the port is not found,
+    /// a default is server is created.
+    pub fn pod_server_rx(
+        &mut self,
+        namespace: &str,
+        pod: &str,
+        port: u16,
+    ) -> Result<watch::Receiver<InboundServer>> {
+        let ns = self
+            .namespaces
+            .by_ns
+            .get_mut(namespace)
+            .ok_or_else(|| anyhow::anyhow!("namespace not found: {}", namespace))?;
+        let pod = ns
+            .pods
+            .by_name
+            .get_mut(pod)
+            .ok_or_else(|| anyhow::anyhow!("pod {}.{} not found", pod, namespace))?;
+        Ok(pod
+            .port_server_or_default(port, &self.cluster_info)
+            .rx
+            .clone())
+    }
+}
+
+impl kubert::index::IndexNamespacedResource<k8s::Pod> for Index {
+    #[instrument(skip_all, name = "pod.apply", fields(
+        ns = %pod.namespace().unwrap(),
+        pod = %pod.name(),
+    ))]
+    fn apply(&mut self, pod: k8s::Pod) {
+        let namespace = pod.namespace().unwrap();
+        let name = pod.name();
+        let port_names = pod::tcp_port_names(pod.spec);
+        let meta = pod::Meta::from_metadata(pod.metadata);
+
+        // Add or update the pod. If the pod was not already present in the
+        // index with the same metadata, index it against the policy resources,
+        // updating its watches.
+        let ns = self.namespaces.get_or_default(namespace);
+        match ns.pods.update(name, meta, port_names) {
+            Ok(None) => {}
+            Ok(Some(pod)) => pod.reindex_servers(&ns.policy),
+            Err(error) => {
+                tracing::error!(%error, "Illegal pod update");
+            }
+        }
+    }
+
+    #[instrument(skip(self), name = "srv.delete")]
```

Comment:
this is the impl for `Pod`, shouldn't this say
```suggestion
    #[instrument(skip(self), name = "pod.delete")]
```

Reply:
also, would we want to include the ns and pod name here?
```suggestion
    #[instrument(skip(self), name = "pod.delete", fields(ns = %namespace, pod = %pod)]
```